README.md
webcache [![Build Status](https://secure.travis-ci.org/fengmk2/webcache.png)](http://travis-ci.org/fengmk2/webcache)
=======

![logo](https://raw.github.com/fengmk2/webcache/master/logo.png)

Web Cache middleware base on `req.url`.

`webcache` can cache all `'text/*', 'application/javascript', 'application/json'` content type.

You can use any cache store engine which implement `set()` and `get()` methods.

* jscoverage: [98%](http://fengmk2.github.com/coverage/webcache.html)

## Support Cache Store

* [redis](https://github.com/mranney/node_redis)
* [mredis](https://github.com/dead-horse/multi_redis)
* [tair](https://github.com/sunfang1cn/node-tair) (TODO)

## Install

```bash
$ npm install webcache
```

## Usage

```js
var connect = require('connect');
var webcache = require('webcache');
var redis = require('redis').createClient();
var app = connect();

app.use(webcache{
  webcache.createStore('redis', redis),
  [
    // cache `GET /article/*` response for one hour, ignore querystring params, enable browser cache
    { match: /^\/article\/\w+/, maxAge: 3600000, ignoreQuerystring: true, clientCache: true },
    // cache `GET /` one day
    { match: /^\/$/, maxAge: 3600000 * 24 },
    // cache `GET /commnet[s]` 
    { match: /^\/comments?/ }, // 5 minutes cache
  ],
  { version: 2012 }
));

// ... ignore ...
```

## Authors

```bash
$ git summary 

 project  : webcache
 repo age : 6 months
 active   : 4 days
 commits  : 15
 files    : 13
 authors  : 
    13  fengmk2                 86.7%
     2  sunfang1cn              13.3%
```

## License 

(The MIT License)

Copyright (c) 2012 - 2013 fengmk2 &lt;fengmk2@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.0.5 / 2013-05-14 
==================

  * fixed test cases
  * use blanket instead of webcache
  * Merge pull request #2 from sunfang1cn/master
  * fix issues for https://github.com/fengmk2/webcache/pull/2
  * add ignore one or some query param key options

0.0.4 / 2012-11-10 
==================

  * rule.match test on url.pathname

0.0.3 / 2012-11-10 
==================

  * add RedisStore support webcache.createStore("redis", redisClient)
  * add redis test cases

0.0.2 / 2012-11-10 
==================

  * add client cache support

0.0.1 / 2012-11-10 
==================

  * first release


test/app.js
/*!
 * webcache - test/app.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var urlparse = require('url').parse;
var connect = require('connect');
var webcache = require('../');
var app = connect();

app._store = webcache.MemoryStore();

app.use(webcache(
  app._store,
  [
    { match: /^\/article\/\w+/, maxAge: 3600000, ignoreQuerystring: true, clientCache: true },
    // cache `GET /` one day
    { match: /^\/$/, maxAge: 100 },
    // cache `GET /commnet[s]` 
    { match: /^\/comments?/ }, // 5 minutes cache
  ],
  { version: 2012 }
));

app.use(function (req, res, next) {
  if (req.url.indexOf('/error') >= 0) {
    res.statusCode = 500;
  }

  if (req.url.indexOf('/404') >= 0) {
    res.statusCode = 404;
  }

  var info = urlparse(req.url, true);
  info.query = info.query || {};
  var contentType = info.query.content_type || 'text/html';
  res.setHeader('Content-Type', contentType);
  if (info.query.nocache) {
    res.setHeader('Cache-Control', 'no-cache');
  }
  res.end(req.method + ' ' + req.url);
});

module.exports = app;

test/server.js
/*!
 * webcache - test/app.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var urlparse = require('url').parse;
var connect = require('connect');
var webcache = require('../');
var path = require('path');
var fs = require('fs');

var textpath = path.join(__dirname, 'webcache.test.js');
var LARGE_TEXT = fs.readFileSync(textpath);
var IMAGE = fs.readFileSync(path.join(path.dirname(__dirname), 'logo.png'));

exports.create = function (store) {
  var app = connect();
  app._store = store;

  app.use(webcache(
    app._store,
    [
      { match: /^\/article\/\w+/, maxAge: 3600000, ignoreQuerystring: true, clientCache: true },
      // cache `GET /` one day
      { match: /^\/$/, maxAge: 1000, ignoreQuerystring: true },
      // cache `GET /commnet[s]` 
      { match: /^\/comments?/ }, // 5 minutes cache
    ],
    { version: 2012, ignoreQueryParams: ['spm', 'sp2'] }
  ));

  app.use(function (req, res, next) {
    if (req.url.indexOf('/error') >= 0) {
      res.statusCode = 500;
    }

    if (req.url.indexOf('/404') >= 0) {
      res.statusCode = 404;
    }

    var info = urlparse(req.url, true);
    info.query = info.query || {};
    var contentType = info.query.content_type || 'text/html';
    res.setHeader('Content-Type', contentType);
    if (info.query.nocache) {
      res.setHeader('Cache-Control', 'no-cache');
    }

    if (req.url === '/article/large') {
      return res.end(LARGE_TEXT); 
    }

    if (req.url === '/article/image') {
      res.setHeader('Content-Type', 'image/png');
      return res.end(IMAGE);
    }

    res.end(req.method + ' ' + req.url);
  });
  return app;
};


test/webcache.test.js
/*!
 * webcache - test/webcache.test.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var fs = require('fs');
var webcache = require('../');
var pedding = require('pedding');
var should = require('should');
var request = require('supertest');
var server = require('./server');
var redis = require('redis');
var mredis = require('mredis');
var interceptor = require('interceptor');

var mockRedisServer = interceptor.create('localhost:6379');
mockRedisServer.listen(63801);
var mredisClient = mredis.createClient({
  server: [ 'localhost:6379', 'localhost:63801' ]
});
var redisClient = redis.createClient();
var version = 'WebCache' + require('../package.json').version;

function cacheKey(key) {
  return 'wc_' + key + '_2012';
}

describe('webcache.test.js', function () {

  var stores = [
    [ 'MemoryStore', webcache.MemoryStore() ],
    [ 'RedisStore', webcache.createStore('redis', redisClient) ],
    [ 'MRedis', webcache.createStore('redis', mredisClient) ],
  ];

  stores.forEach(function (item) {

    describe(item[0], function () {

      var store = item[1];
      var cleanupCache = function (callback) {
        var keys = [
          '/article/foo',
          '/comments/foobar',
          '/article/image',
          '/article/json',
          '/article/javascript',
          '/article/large',
          '/comments/foobar?foo=bar',
          '/'
        ];
        callback = pedding(keys.length, callback);
        keys.forEach(function (key) {
          store.set(cacheKey(key), null, null, callback);
        });
      };
      
      var app = server.create(store);
      var _get = store.get;
      beforeEach(function (done) {
        store.get = function (key, callback) {
          if (key.indexOf('mock_get_error') >= 0) {
            process.nextTick(function () {
              callback(new Error('mock get ' + key + ' error'));
            });
            return;
          }
          _get.call(this, key, callback);
        };
        cleanupCache(done);
      });
      afterEach(function (done) {
        store.get = _get;
        cleanupCache(done);
      });

      it('should throw TypeError when params wrong', function () {
        (function () {
          webcache();
        }).should.throw('cache must support #get() and #set()');
        (function () {
          webcache({});
        }).should.throw('cache must support #get() and #set()');
        (function () {
          webcache(app._store);
        }).should.throw('rules must not empty');
        (function () {
          webcache(app._store, []);
        }).should.throw('rules must not empty');
      });

      it('should cache GET /article/foo', function (done) {
        done = pedding(2, done);

        request(app)
        .get('/article/foo')
        .expect('GET /article/foo')
        .expect('Content-Type', 'text/html')
        .expect(200, function (err, res) {
          should.not.exist(err);
          res.headers.should.not.have.property('X-Cache-By');
          request(app)
          .get('/article/foo')
          .expect('X-Cache-By', version)
          .expect('Cache-Control', 'public, max-age=3600')
          .expect('Content-Type', 'text/html')
          .expect('GET /article/foo')
          .expect(200, done);

          request(app)
          .get('/article/foo?t=123123')
          .expect('X-Cache-By', version)
          .expect('Cache-Control', 'public, max-age=3600')
          .expect('Content-Type', 'text/html')
          .expect('GET /article/foo')
          .expect(200, done);
        });
      });

      it('should cache GET /article/large', function (done) {
        done = pedding(2, done);

        var content = fs.readFileSync(__filename, 'utf8');

        request(app)
        .get('/article/large')
        .expect(content)
        .expect('Content-Type', 'text/html')
        .expect(200, function (err, res) {
          should.not.exist(err);
          res.headers.should.not.have.property('X-Cache-By');
          request(app)
          .get('/article/large')
          .expect('X-Cache-By', version)
          .expect('Cache-Control', 'public, max-age=3600')
          .expect('Content-Type', 'text/html')
          .expect(content)
          .expect(200, done);

          request(app)
          .get('/article/large?t=123123')
          .expect('X-Cache-By', version)
          .expect('Cache-Control', 'public, max-age=3600')
          .expect('Content-Type', 'text/html')
          .expect(content)
          .expect(200, done);
        });
      });

      it('should cache GET /comments/foobar and not cache /comments/foobar?foo=bar', function (done) {
        done = pedding(2, done);

        request(app)
        .get('/comments/foobar')
        .expect('GET /comments/foobar')
        .expect('Content-Type', 'text/html')
        .expect(200, function (err, res) {
          should.not.exist(err);
          res.headers.should.not.have.property('x-cache-by');
          request(app)
          .get('/comments/foobar')
          .expect('X-Cache-By', version)
          .expect('Content-Type', 'text/html')
          .expect('GET /comments/foobar')
          .expect(200, function (err, res) {
            should.not.exist(err);
            done();
          });

          request(app)
          .get('/comments/foobar?foo=bar')
          .expect('Content-Type', 'text/html')
          .expect('GET /comments/foobar?foo=bar')
          .expect(200, function (err, res) {
            should.not.exist(err);
            res.headers.should.not.have.property('x-cache-by');
            done();
          });
        });
      });

      it('should cache / 1000ms', function (done) {
        request(app)
        .get('/')
        .expect('GET /')
        .expect(200, function (err, res) {
          request(app)
          .get('/')
          .expect('X-Cache-By', version)
          .expect('GET /')
          .expect(200, function (err, res) {
            should.not.exist(err);
            res.headers.should.not.have.property('Cache-Control');
            setTimeout(function () {
              request(app)
              .get('/')
              .expect('GET /')
              .expect(200, function (err, res) {
                should.not.exist(err);
                res.headers.should.not.have.property('x-cache-by');
                done();
              });
            }, 2000);
          });
        });
      });

      it('should cache / and ignore querystring effect', function (done) {
        request(app)
        .get('/')
        .expect('GET /')
        .expect(200, function (err, res) {
          request(app)
          .get('/?foo=bar')
          .expect('X-Cache-By', version)
          .expect('GET /')
          .expect(200, done);
        });
      });
   
      it('should cache /comments and ignore ?spm=xxxx query params', function (done) {
        request(app)
        .get('/comments?spm=2.3.4.5&app=123')
        .expect(200, function (err, res) {
          request(app)
          .get('/comments?app=123&spm=1.2.3.4')
          .expect('X-Cache-By', version)
          .expect(200, done);
        });
      });
      
      it('should cache /comments and ignore ?sp2=xxxx query params', function (done) {
        request(app)
        .get('/comments?sp2=test&app=1243&sc=cdg2')
        .expect(200, function (err, res) {
          request(app)
          .get('/comments?app=1243&sc=cdg2&sp2=12.3.4')
          .expect('X-Cache-By', version)
          .expect(200, done);
        });
      });

      it('should cache /article/javascript?content_type=application/javascript', function (done) {
        request(app)
        .get('/article/javascript?content_type=application/javascript')
        .expect('Content-Type', 'application/javascript')
        .expect(200, function (err, res) {
          should.not.exist(err);
          res.text.should.equal('GET /article/javascript?content_type=application%2Fjavascript');
          
          request(app)
          .get('/article/javascript?content_type=application/javascript')
          .expect('X-Cache-By', version)
          .expect('Content-Type', 'application/javascript')
          .expect(200, function (err, res) {
            should.not.exist(err);
            res.text.should.equal('GET /article/javascript?content_type=application%2Fjavascript');
            done();
          });
        });
      });

      it('should cache /article/json?content_type=application/json', function (done) {
        request(app)
        .get('/article/json?content_type=application/json')
        .expect('GET /article/json?content_type=application%2Fjson')
        .expect('Content-Type', 'application/json')
        .expect(200, function (err, res) {
          should.not.exist(err);
          res.headers.should.not.have.property('x-cache-by');
          request(app)
          .get('/article/json?content_type=application/json')
          .expect('GET /article/json?content_type=application%2Fjson')
          .expect('Content-Type', 'application/json')
          .expect('X-Cache-By', version)
          .expect(200, done);
        });
      });

      describe('no cache', function () {
        it('should not cache when not match any rules', function (done) {
          request(app)
          .get('/foo')
          .expect('GET /foo')
          .expect(200, function (err, res) {
            should.not.exist(err);
            request(app)
            .get('/foo')
            .expect('GET /foo')
            .expect(200, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              done();
            });
          });
        });

        it('should not cache when request with no-cache headers', function (done) {
          request(app)
          .get('/comments')
          .expect('GET /comments')
          .expect(200, function (err, res) {
            should.not.exist(err);
            request(app)
            .get('/comments')
            .set('Cache-Control', 'no-cache')
            .expect('GET /comments')
            .expect(200, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              done();
            });
          });
        });

        it('should not cache when response error', function (done) {
          request(app)
          .get('/article/error')
          .expect('GET /article/error')
          .expect(500, function (err, res) {
            should.not.exist(err);
            res.headers.should.not.have.property('x-cache-by');
            request(app)
            .get('/article/error')
            .expect('GET /article/error')
            .expect(500, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              done();
            });
          });
        });

        it('should not cache when response 404', function (done) {
          request(app)
          .get('/article/404')
          .expect('GET /article/404')
          .expect(404, function (err, res) {
            should.not.exist(err);
            res.headers.should.not.have.property('x-cache-by');
            request(app)
            .get('/article/404')
            .expect('GET /article/404')
            .expect(404, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              done();
            });
          });
        });

        it('should not cache when response no-cache header', function (done) {
          request(app)
          .get('/article/foo?nocache=1')
          .expect('GET /article/foo?nocache=1')
          .expect(200, function (err, res) {
            should.not.exist(err);
            res.headers.should.not.have.property('x-cache-by');
            request(app)
            .get('/article/foo?nocache=1')
            .expect('GET /article/foo?nocache=1')
            .expect(200, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              done();
            });
          });
        });

        it('should not cache when cache store get() error', function (done) {
          request(app)
          .get('/comments/mock_get_error')
          .expect('GET /comments/mock_get_error')
          .expect(200, function (err, res) {
            should.not.exist(err);
            res.headers.should.not.have.property('x-cache-by');
            request(app)
            .get('/comments/mock_get_error')
            .expect('GET /comments/mock_get_error')
            .expect(200, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              app._store.get = _get;
              request(app)
              .get('/comments/mock_get_error')
              .expect('GET /comments/mock_get_error')
              .expect(200, function (err, res) {
                should.not.exist(err);
                res.headers.should.have.property('x-cache-by');
                done();
              });
            });
          });
        });

        it('should not cache POST request', function (done) {
          request(app)
          .post('/')
          .expect('POST /')
          .expect(200, function (err, res) {
            should.not.exist(err);
            request(app)
            .post('/')
            .expect('POST /')
            .expect(200, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              done(err);
            });
          });
        });

        it('should not cache DELETE request', function (done) {
          request(app)
          .del('/')
          .expect('DELETE /')
          .expect(200, function (err, res) {
            should.not.exist(err);
            request(app)
            .del('/')
            .expect('DELETE /')
            .expect(200, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              done(err);
            });
          });
        });

        it('should not cache PUT request', function (done) {
          request(app)
          .put('/')
          .expect('PUT /')
          .expect(200, function (err, res) {
            should.not.exist(err);
            request(app)
            .put('/')
            .expect('PUT /')
            .expect(200, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              done(err);
            });
          });
        });

        it('should not cache /article/image request', function (done) {
          request(app)
          .get('/article/image')
          .expect(200, function (err, res) {
            should.not.exist(err);
            res.headers.should.not.have.property('x-cache-by');
            request(app)
            .get('/article/image')
            .expect(200, function (err, res) {
              should.not.exist(err);
              res.headers.should.not.have.property('x-cache-by');
              done(err);
            });
          });
        });

      });
    });
  });

  it('should createStore() throw error when cache type not support', function () {
    (function () {
      webcache.createStore('tair', {});
    }).should.throw('tair not support, please implement the get() and set() methods yourself');
  });
  
});



lib/webcache.js
/*!
 * webcache - lib/webcache.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

require('buffer-concat');
var debug = require('debug')('webcache');
var eventproxy = require('eventproxy');
var urlparse = require('url').parse;

/**
 * Parse the given Cache-Control `str`.
 *
 * @param {String} str
 * @return {Object}
 * @api private
 */
var parseCacheControl = function (str) {
  var directives = str.split(',');
  var obj = {};

  for (var i = 0, len = directives.length; i < len; i++) {
    var parts = directives[i].split('=');
    var key = parts.shift().trim();
    var val = parseInt(parts.shift(), 10);

    obj[key] = isNaN(val) ? true : val;
  }

  return obj;
};

/**
 * Web Cache middleware.
 *
 * ```js
 * app.use(webcache(
 *   webcache.redisStore(redis),
 *   [
 *     { match: /^\/article\/\w+/, maxAge: 3600000, ignoreQuerystring: true },
 *     { match: /^\/$/, maxAge: 3600000 * 24 },
 *     { match: /^\/comments?/ }, // 5 minutes cache
 *   ],
 *   { version: 2012 }
 * ));
 * ```
 *
 * @param {CacheStore} cache, cache store impl `get()` and `set()` methods.
 *   #get(key, callback) and 
 *   #set(key, value, maxAge, [callback]), or use a array and exports.setCacheIndex
 * @param {Array} rules, match url cache rules.
 *   - {Object} rule: {
 *     - {RegExp} match: regex to detect which `req.url` need to be cache.
 *     - {Number} maxAge: cache millisecond, default is `options.maxAge`.
 *     - {Boolean} ignoreQuerystring: ignore `req.url` querystring params.
 *     - {Boolean} clientCache: client side cache, default is `options.clientCache`.
 *     - {Array} ignoreQueryParams: An Array of query params key name in url.
 *   }
 * @param options
 *   - {Number} maxAge: global cache millisecond, default is 300000ms (5 minutes).
 *   - {String} version: cache version, append to cache store key, default is `''`.
 *   - {Boolean} ignoreQuerystring: ignore `req.url` querystring params, default is false.
 *   - {Boolean} clientCache: client side cache, default is false.
 *   - {Array} ignoreQueryParams: An Array of query params key name in url.
 * @returns {Function (req, res, next)}
 * @api public
 */
exports = module.exports = function webcache(cache, rules, options) {
  if (!cache || typeof cache.get !== 'function' || typeof cache.set !== 'function') {
    throw new TypeError('cache must support #get() and #set()');
  }
  if (!rules || !rules.length) {
    throw new TypeError('rules must not empty');
  }

  options = options || {};
  var defaultMaxAge = options.maxAge || 300000;
  var defaultIgnoreQuerystring = options.ignoreQuerystring || false;
  var defaultIgnoreQueryParams = options.ignoreQueryParams || [];
  var version = options.version || '';
  var clientCache = !!options.clientCache;

  var TEXT_RE = /^(?:text\/|application\/json|application\/javascript)/i;

  for (var i = 0; i < rules.length; i++) {
    var rule = rules[i];
    rule.maxAge = rule.maxAge || defaultMaxAge;
    rule.ignoreQuerystring = rule.ignoreQuerystring || defaultIgnoreQuerystring;
    var ignoreQueryParams = rule.ignoreQueryParams || defaultIgnoreQueryParams;
    rule.ignoreQueryParamsRE = [];
    // like '?spm=2.3.45&' <=> /(\&|\?)spm[^\&]*\&?/
    for (var j = 0; j < ignoreQueryParams.length; j ++) {
      rule.ignoreQueryParamsRE[j] = new RegExp('(\\&|\\?)' + ignoreQueryParams[j] + '[^\\&]*\\&?' , 'ig');
    }
    rule.clientCache = rule.clientCache || clientCache;
  }

  return function webcache(req, res, next) {
    if (req.method !== 'GET') {
      return next();
    }

    var matchRule = null;
    var urlinfo = urlparse(req.url);
    for (var i = 0; i < rules.length; i++) {
      var rule = rules[i];
      if (rule.match.test(urlinfo.pathname)) {
        matchRule = rule;
      }
    }

    if (!matchRule) {
      return next();
    }
    var url = req.url;
    if (matchRule.ignoreQuerystring) {
      url = urlinfo.pathname;
    }
    for (var i = 0; i < matchRule.ignoreQueryParamsRE.length; i ++) {
      url = url.replace(matchRule.ignoreQueryParamsRE[i], function (match0, match1) { 
        // if first char is '?', leave it
        if (match1 === '?') { 
          return match1; 
        }
        return ''; 
      });
    }
    var key = 'wc_' + url + '_' + version;
    var keyContentType = key + '_ct';
    var maxAge = matchRule.maxAge;

    var nextHandle = function () {
      var chunks = [];
      var size = 0;
      res.__source_write = res.write;
      res.write = function (chunk, encoding) {
        this.__source_write(chunk, encoding);

        if (!Buffer.isBuffer(chunk)) {
          chunk = new Buffer(chunk, encoding);
        }
        chunks.push(chunk);
        size += chunk.length;
      };
      res.__source_end = res.end;
      res.end = function (chunk, encoding) {
        if (chunk) {
          this.write(chunk, encoding);
        }
        this.end = this.__source_end;
        this.end();

        if (res.statusCode !== 200) {
          // dont cache non 200 status response
          return;
        }

        var cc = res.getHeader('cache-control');
        if (cc && parseCacheControl(cc)['no-cache']) {
          return;
        }

        if (chunks.length > 0) {
          var buf = Buffer.concat(chunks, size);
          if (buf && buf.length > 0) {
            var contentType = res.getHeader('Content-Type');
            if (contentType) {
              var need = TEXT_RE.test(contentType);
              debug('check %j %s %s', req.url, contentType, need);
              if (need) {
                cache.set(key, buf, maxAge);
                cache.set(keyContentType, contentType, maxAge);
              }
            }
          }
        } 
        
      };
      next();
    };

    var ep = eventproxy.create('content_type', 'content', function (contentType, content) {
      if (content) {
        var cc = req.headers['cache-control'];
        if (!cc || !parseCacheControl(cc)['no-cache']) {
          if (contentType) {
            res.setHeader('Content-Type', contentType);
          }
          res.setHeader('X-Cache-By', 'WebCache' + exports.version);
          if (matchRule.clientCache && maxAge >= 1000) {
            var seconds = parseInt(maxAge / 1000, 10);
            res.setHeader('Cache-Control', 'public, max-age=' + seconds);
          }
          debug('hit %s %s, %d bytes', key, contentType, content.length);
          return res.end(content); 
        }
      }
      debug('miss %s', key);

      nextHandle();
    });
    
    ep.on('error', function (err) {
      debug('[%s] %s', new Date(), err.stack);
      ep.unbind();
      nextHandle();
    });

    cache.get(key, function (err, content) {
      if (err) {
        return ep.emit('error', err);
      }
      ep.emit('content', content);
    });

    cache.get(keyContentType, function (err, keyContentType) {
      if (err) {
        return ep.emit('error', err);
      }
      ep.emit('content_type', keyContentType);
    });
  };
};

/**
 * Cache in process memory, Don't use int production env.
 */
function MemoryStore() {
  if (!(this instanceof MemoryStore)) {
    return new MemoryStore();
  }
  this._data = {};
  if (process.env.NODE_ENV !== 'test') {
    console.warn('[webcache][%s] MUST not use MemoryStore in production env.', new Date());
  } 
}

MemoryStore.prototype.set = function (key, value, maxAge, callback) {
  var expired = 0;
  if (maxAge) {
    expired = Date.now() + maxAge;
  }
  this._data[key] = [value, expired, maxAge];
  process.nextTick(function () {
    callback && callback();
  });
};

MemoryStore.prototype.get = function (key, callback) {
  var data = this._data[key];
  var value = null;
  if (data) {
    if (data[1] && Date.now() >= data[1]) {
      delete this._data[key];
    } else {
      value = data[0];
    }
  }
  process.nextTick(function () {
    callback(null, value);
  });
};

exports.MemoryStore = MemoryStore;

function RedisStore(cache) {
  this.cache = cache;
}

RedisStore.prototype.set = function (key, value, maxAge, callback) {
  if (!value) {
    debug('del %s', key);
    return this.cache.del(key, callback);
  }
  var seconds = 0;
  if (maxAge && maxAge >= 1000) {
    seconds = parseInt(maxAge / 1000, 10);
  }
  debug('set %s %s', key, seconds);
  if (seconds) {
    this.cache.setex(key, seconds, value, callback);
  } else {
    this.cache.set(key, value, callback);
  }
};

RedisStore.prototype.get = function (key, callback) {
  this.cache.get(key, callback);
};

/**
 * Create a cache store.
 * @param {String} type, cache type, e.g.: 'redis', 'tair'
 * @param {Object} cache, cache instance
 * @return {Object} store instance
 */
exports.createStore = function (type, cache) {
  if (type === 'redis') {
    return new RedisStore(cache);
  }
  throw new TypeError(type + ' not support, please implement the get() and set() methods yourself');
};

exports.version = require('../package.json').version;

